homeassistant/components/mqtt/cover.py

```diff
@@ -347,13 +339,44 @@ def position_message_received(msg):
                 template = self._config.get(CONF_VALUE_TEMPLATE)
 
             if template is not None:
-                payload = template.async_render_with_possible_json_value(payload)
-
-            if not payload:
-                _LOGGER.debug("Ignoring empty position message from '%s'", msg.topic)
-                return
+                variables = {
+                    "entity_id": self.entity_id,
+                    "position_open": self._config[CONF_POSITION_OPEN],
+                    "position_closed": self._config[CONF_POSITION_CLOSED],
+                    "tilt_min": self._config[CONF_TILT_MIN],
+                    "tilt_max": self._config[CONF_TILT_MAX],
+                }
+                payload = template.async_render_with_possible_json_value(
+                    payload, variables=variables
+                )
 
-            if payload.isnumeric():
+                if not payload:
+                    _LOGGER.debug(
+                        "Ignoring empty position message from '%s'", msg.topic
+                    )
+                    return
+
+                try:
+                    payload = json_loads(payload)
+                except TypeError:
+                    pass
+                except JSONDecodeError:
+                    pass
+
+                if isinstance(payload, dict):
+                    if "position" not in payload:
+                        _LOGGER.warning(
+                            "Template (position_template) returned JSON without position attribute"
+                        )
+                        return
+                    if "tilt_position" in payload:
+                        if not self._config.get(CONF_TILT_STATE_OPTIMISTIC):
+                            # reset forced set tilt optimistic
+                            self._tilt_optimistic = False
+                        self.tilt_payload_received(payload["tilt_position"])
+                    payload = payload["position"]
+
+            if type(payload) in [int, float] or payload.isnumeric():
```

Comment:
Instead of this plus the `float()` conversion below, maybe like this:
```py
try:
    percentage_payload = self.find_percentage_in_range(
        float(payload), COVER_PAYLOAD
    )
    ...
except ValueError:
    _LOGGER.warning("Payload '%s' is not numeric", payload)
    return